src/core/judgingthread.cpp

```diff
@@ -96,378 +96,210 @@ auto JudgingThread::getNeedRejudge() const -> bool { return needRejudge; }
 
 void JudgingThread::stopJudgingSlot() { stopJudging = true; }
 
+class BufferedStreamReader {
+
+  public:
+	static const int BUFFER_SIZE = 1 << 18; // 128 KiB
+	BufferedStreamReader(const char *filename) : file(fopen(filename, "r")) {
+		result.reserve(32);
+		bufPos = buffer + BUFFER_SIZE;
+		bufEnd = buffer + BUFFER_SIZE;
+		lineNumber = 1;
+		isEof = 0;
+	}
+	~BufferedStreamReader() { fclose(file); }
+
+  private:
+	FILE *file;
+	char buffer[BUFFER_SIZE], *bufEnd, *bufPos;
+	bool isEof;
+	int lineNumber;
+	std::string result; // Fixed Length String,Optimize for speed
+	auto peekChar() -> char;
+	void tryNextLine();
+
+  public:
+	auto valid() const -> bool { return file != nullptr; }
+	auto eof() const -> bool { return isEof; }
+	auto line() const -> int { return lineNumber; }
+	auto nextUntilNewLine() -> std::string &;
+	auto nextUntilSpace() -> std::string &;
+};
+
+auto BufferedStreamReader::peekChar() -> char {
+	if (isEof) {
+		return '\0';
+	}
+	if (bufPos == bufEnd) {
+		bufPos = buffer;
+		bufEnd = buffer + fread(buffer, 1, BUFFER_SIZE, file);
+		if (bufPos == bufEnd) {
+			isEof = true;
+			return '\0';
+		}
+	}
+	return *bufPos;
+}
+
+void BufferedStreamReader::tryNextLine() {
+	char c = peekChar();
+	if (c != '\r' && c != '\n')
+		return;
+	if (c == '\r')
+		bufPos++;
+	c = peekChar();
+	if (c == '\n')
+		bufPos++;
+	lineNumber++;
+}
+
+auto BufferedStreamReader::nextUntilNewLine() -> std::string & {
```

Comment:
since we use it as `const std::string&`, maybe `std::string_view` is a better choice
Also, since we use `QString` in all code, `QString` and `QStringView` may be a better choice for consistency.

Reply:
确实